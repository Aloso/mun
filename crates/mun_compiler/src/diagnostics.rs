use mun_hir::diagnostics::{Diagnostic as HirDiagnostic, DiagnosticSink};
use mun_hir::{FileId, HirDatabase, HirDisplay, Module};
use mun_syntax::{ast, AstNode, SyntaxKind};
use std::cell::RefCell;

mod emit;

pub use emit::Emit;
use mun_errors::{Diagnostic, Level};

/// Constructs diagnostic messages for the given file.
pub fn diagnostics(db: &impl HirDatabase, file_id: FileId) -> Vec<Diagnostic> {
    let parse = db.parse(file_id);
    let mut result = Vec::new();

    result.extend(parse.errors().iter().map(|err| Diagnostic {
        level: Level::Error,
        loc: err.location(),
        message: format!("Syntax Error: {}", err),
    }));

    let result = RefCell::new(result);
    let mut sink = DiagnosticSink::new(|d| {
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: d.highlight_range().into(),
            message: d.message(),
        });
    })
    .on::<mun_hir::diagnostics::UnresolvedValue, _>(|d| {
        let text = d.expr.to_node(&parse.syntax_node()).text().to_string();
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: d.highlight_range().into(),
            message: format!("could not find value `{}` in this scope", text),
        });
    })
    .on::<mun_hir::diagnostics::UnresolvedType, _>(|d| {
        let text = d
            .type_ref
            .to_node(&parse.syntax_node())
            .syntax()
            .text()
            .to_string();
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: d.highlight_range().into(),
            message: format!("could not find type `{}` in this scope", text),
        });
    })
    .on::<mun_hir::diagnostics::ExpectedFunction, _>(|d| {
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: d.highlight_range().into(),
            message: format!("expected function, found `{}`", d.found.display(db)),
        });
    })
    .on::<mun_hir::diagnostics::MismatchedType, _>(|d| {
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: d.highlight_range().into(),
            message: format!(
                "expected `{}`, found `{}`",
                d.expected.display(db),
                d.found.display(db)
            ),
        });
    })
    .on::<mun_hir::diagnostics::DuplicateDefinition, _>(|d| {
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: match d.definition.kind() {
                SyntaxKind::FUNCTION_DEF => {
                    ast::FunctionDef::cast(d.definition.to_node(&parse.syntax_node()))
                        .map(|f| f.signature_range())
                        .unwrap_or_else(|| d.highlight_range())
                        .into()
                }
                _ => d.highlight_range().into(),
            },
            message: d.message(),
        });
    })
    .on::<mun_hir::diagnostics::PossiblyUninitializedVariable, _>(|d| {
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: d.highlight_range().into(),
            message: format!(
                "use of possibly-uninitialized variable: `{}`",
                d.pat.to_node(&parse.syntax_node()).text().to_string()
            ),
        })
    })
    .on::<mun_hir::diagnostics::AccessUnknownField, _>(|d| {
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: ast::FieldExpr::cast(d.expr.to_node(&parse.syntax_node()))
                .map(|f| f.field_range())
                .unwrap_or_else(|| d.highlight_range())
                .into(),
            message: format!(
                "no field `{}` on type `{}`",
                d.name,
                d.receiver_ty.display(db),
            ),
        })
    });

    Module::from(file_id).diagnostics(db, &mut sink);

    drop(sink);
    result.into_inner()
}
