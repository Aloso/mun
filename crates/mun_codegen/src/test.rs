use crate::{mock::MockDatabase, IrDatabase};
use mun_hir::diagnostics::DiagnosticSink;
use mun_hir::Module;
use std::cell::RefCell;
use std::path::PathBuf;
use test_utils::{dir_tests, project_dir};

fn test_data_dir() -> PathBuf {
    project_dir().join("crates/mun_codegen/tests/data/")
}

#[test]
fn ir_tests() {
    dir_tests(&test_data_dir(), &["ir"], |text, _path| {
        let (db, file_id) = MockDatabase::with_single_file(text);

        let messages = RefCell::new(Vec::new());
        let mut sink = DiagnosticSink::new(|diag| {
            messages.borrow_mut().push(diag.message());
        });
        if let Some(module) = Module::package_modules(&db)
            .iter()
            .find(|m| m.file_id() == file_id)
        {
            module.diagnostics(&db, &mut sink)
        }
        drop(sink);
        let messages = messages.into_inner();

        if !messages.is_empty() {
            messages.join("\n")
        } else {
            format!(
                "{}",
                db.module_ir(file_id)
                    .llvm_module
                    .print_to_string()
                    .to_string()
            )
        }
    });
}
