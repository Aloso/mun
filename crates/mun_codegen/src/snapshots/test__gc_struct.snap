---
source: crates/mun_codegen/src/test.rs
expression: "struct(gc) Foo { a: int, b: int };\n\nfn foo() {\n    let a = Foo { a: 3, b: 4 };\n    a.b += 3;\n    let b = a;\n}"
---
; ModuleID = 'main.mun'
source_filename = "main.mun"

%Foo = type { i64, i64 }

define void @foo() {
body:
  %b4 = alloca %Foo*
  %a = alloca %Foo*
  %malloccall = tail call i8* @malloc(i32 trunc (i64 mul nuw (i64 ptrtoint (i64* getelementptr (i64, i64* null, i32 1) to i64), i64 2) to i32))
  %Foo = bitcast i8* %malloccall to %Foo*
  store %Foo { i64 3, i64 4 }, %Foo* %Foo
  store %Foo* %Foo, %Foo** %a
  %deref = load %Foo*, %Foo** %a
  %Foo.b = getelementptr inbounds %Foo, %Foo* %deref, i32 0, i32 1
  %b = load i64, i64* %Foo.b
  %add = add i64 %b, 3
  %deref1 = load %Foo*, %Foo** %a
  %Foo.b2 = getelementptr inbounds %Foo, %Foo* %deref1, i32 0, i32 1
  store i64 %add, i64* %Foo.b2
  %a3 = load %Foo*, %Foo** %a
  store %Foo* %a3, %Foo** %b4
  ret void
}

declare noalias i8* @malloc(i32)

